ENH: Use `highspy` in `linprog`

HaoZeke opened the pull request:
Supersedes https://github.com/scipy/scipy/pull/18642/.

#### Reference issue
Closes https://github.com/scipy/scipy/issues/15915. Closes #15888. Closes #19734.

#### What does this implement/fix?
- Replaces existing `linprog` Cython bindings with upstream `highspy`

Tests pass (locally anyway). However, there are still plenty of improvements which can be made (once `highspy` works a bit more robustly https://github.com/ERGO-Code/HiGHS/pull/1405)
#### Additional information
Needs a PR over at `highs` (https://github.com/ERGO-Code/HiGHS/pull/1460) and over at the SciPy fork (https://github.com/scipy/HiGHS/pull/63).


Draft status until:
- ~[ ] Changes merged upstream in a new PyPI release~ Unrelated
- [x] Be 100 % backwards compatible (or document changes)
- [x] Fixup the enum translation (should be quick, just too tired rn)
- [x] Fixup the commit history, early stages are derived from #18642 and should have co-commit credit for @mckib2
- [x] Figure out / fix the xfail on Linux 32-bit
- [x] Add the git submodule back
- [x] Point to the `scipy` fork of `highs`

HaoZeke commented:
Also @mckib2 if you'd like to take a look that'd be great (or maybe after its out of draft status).

rgommers commented on scipy/optimize/_highs/meson.build:
```diff
@@ -1,284 +1,300 @@
-highs_define_macros = [
```
Useful to just delete all these things rather than comment them out?

rgommers reviewed (commented):
What's here looks good - lots of complexity gets removed. Hard to say more because an essential bit - actually building `highspy` as part of the SciPy build - seems to be missing.

HaoZeke commented:
This is almost done, except the build process in SciPy is stricter than over at HiGHs (w.r.t warnings) so some more work is required.

rgommers commented:
You can silence classes of warnings by adding defining them (if they're not already present) here: https://github.com/scipy/scipy/blob/21dcad26e80a6f5dbda4ef7827cad6c2ecb20291/scipy/meson.build#L287-L300

and then adding them to `cpp_args` for the relevant build target.

rgommers commented:
It'd still be nice to get them fixed upstream, but that shouldn't be blocking for finalizing this PR.

HaoZeke commented:
> It'd still be nice to get them fixed upstream, but that shouldn't be blocking for finalizing this PR.

Thanks that helped, but I'm not sure of the best approach for the rest, it seems like using a shared library ends up with (expected) lookup errors, but the static library fails the wheel build. In both cases there's a too many public symbols error. What is the best approach to dealing with libraries which (ideally) should be installed by the user? I guess I'm missing something about the way the static library should be linked, since other sub-libraries are linked statically (`quadpack`).

rgommers commented:
The current version seems quite close to ready, modulo the use of a wrap file (I assume that's for convenience during development, and it'll change to a git submodule at the end) and this error:
```
meson-python: error: Could not map installation path to an equivalent wheel directory: '{libdir_static}/libhighs.a'
```

The static library linking looks fine here, the only problem is that the `libhighs` definition in the subproject has `install: true` and hence by default it's in the list of to-be-installed files, but that's not what we want. This fixes the wheel build:
```
python -m build -Cinstall-args=--skip-subprojects
```

To make that the default behavior, add this to `pyproject.toml`:
```toml
[tool.meson-python.args]
install = ['--skip-subprojects']
```

HaoZeke commented:
Thanks @rgommers, I'll get to updating this in a few days but:

> ... modulo the use of a wrap file (I assume that's for convenience during development, and it'll change to a git submodule at the end) 

I was thinking we ought to keep it as a submodule? It should change to point to the Scipy/Highs instead of my own but other than that it would make more sense to keep the meson-first subproject approach. The only issue I can think of is that the `highs` repo ends up [under subprojects](https://mesonbuild.com/Subprojects.html#why-must-all-subprojects-be-inside-a-single-directory) which may be logically less than pleasant.

The way the build is structured now, the subproject defines all the necessary variables (including the files needed), which means we would very rarely need to update the `meson.build` in SciPy. e.g. With the submodule system, if a new file is added, we need to update the `meson.build` in SciPy. Under the current scheme, we'd continue as-is because it would be part of the variables exported from the subproject (we'd still need to update the wrap-file hash though).

Perhaps @eli-schwartz has more comments on submodules v/s subprojects?

eli-schwartz commented:
Using it as a subproject is very much preferable. You get component isolation: it's not SciPy's job to be concerned about how highs gets built, you just want to automatically use whatever the highs subproject does to define its static-library artifact(s), and then you consume them.

This also unlocks the ability to switch from:
```meson
highs_proj = subproject('highs', ......
```
to
```meson
highs_proj = dependency('highs')
```

along with specifying in *SciPy*, `default_options: ['wrap_mode=forcefallback']`

This would allow downstream builders (especially linux distros) the option to build highs as an external package and then build SciPy against that. But people doing `pip install scipy` or using `python dev.py build` would by default use the subproject.

eli-schwartz commented:
`meson dist --include-subprojects` will create sdists that have the subproject included from the wrap file.

Whether to use subprojects via a wrap file or a git submodule (or both at the same time!) is largely unimportant as far as meson is concerned. The main thing that strikes me about the topic is that git submodules tend to be larger to download than a tarball, and, if you create "light" sdists that only download the subprojects when the person building the wheel hasn't disabled the use of subprojects, the tarball version works without git installed, while the submodule requires installing git.

I think both options are defensible choices, but adding a meson subproject as a git submodule will probably be more pleasant for you if you expect most people to interact with the source code via git using long-running clones, distribute sdists with subprojects included, and don't support GitHub autogenerated tag tarballs at all...

... which are all qualities that SciPy has.

rgommers commented:
Thanks, that explains it. Then I had the right thing in mind:

> modulo the use of a wrap file (I assume that's for convenience during development, and it'll change to a git submodule at the end)

The wrap file is indeed nicer now, when you're developing this PR while making fixes to HiGHS as well. And once it's done, you can swap the wrap file for a git submodule under `subprojects/`.

HaoZeke commented:
Failures seem unrelated (`cython_special`) but I can reproduce them (with a newly created environment) so I assume it has something to do with a dependency update.

HaoZeke commented on scipy/optimize/_highs/cython/src/_highs_wrapper.pyx:
```diff
@@ -1,736 +0,0 @@
-        return {
-            'status': <int> model_status,
-            'message': f'model_status is {highs.modelStatusToString(model_status).decode()}; '
-                       f'primal_status is {utilBasisStatusToString(<HighsBasisStatus> info.primal_solution_status).decode()}',
```
This seems to have been a bug to begin with, for example (using the documentation):
```python
import numpy as np
from scipy.optimize import linprog

# Objective function vector
c = np.array([-29.0, -45.0, 0.0, 0.0])

# Inequality constraints matrix
A_ub = np.array([
    [1.0, -1.0, -3.0, 0.0],
    [-2.0, 3.0, 7.0, -3.0]
])
# Inequality constraints vector
b_ub = np.array([5.0, -10.0])

# Equality constraints matrix
A_eq = np.array([
    [2.0, 8.0, 1.0, 0.0],
    [4.0, 4.0, 0.0, 1.0]
])
# Equality constraints vector
b_eq = np.array([60.0, 60.0])

# Bounds for each variable
x0_bounds = (0, None)
x1_bounds = (0, 5.0)
x2_bounds = (-np.inf, 0.5)  # +/- np.inf can be used instead of None
x3_bounds = (-3.0, None)
bounds = [x0_bounds, x1_bounds, x2_bounds, x3_bounds]

# Linear programming solution
result = linprog(c, A_ub=A_ub, b_ub=b_ub, A_eq=A_eq, b_eq=b_eq, bounds=bounds)

# Display message from the optimization result
print(result.message)
```
Which would previously report:
`The problem is infeasible. (HiGHS Status 8: model_status is Infeasible; primal_status is At lower/fixed bound)`

However, this is actually not true, and forcing the primal_solution_status instead of using `highs.solutionStatusToString(info.primal_solution_status)` shouldn't be happening. The following (by inserting a breakpoint):

```python
(Pdb) basis.row_status
[<HighsBasisStatus.kUpper: 2>, <HighsBasisStatus.kUpper: 2>, <HighsBasisStatus.kLower: 0>, <HighsBasisStatus.kLower: 0>]
(Pdb) basis.col_status
[<HighsBasisStatus.kBasic: 1>, <HighsBasisStatus.kBasic: 1>, <HighsBasisStatus.kBasic: 1>, <HighsBasisStatus.kLower: 0>]
(Pdb) [_h.Highs_().basisStatusToString(x) for x in basis.col_status]
['Basic', 'Basic', 'Basic', 'At lower/fixed bound']
(Pdb) [_h.Highs_().basisStatusToString(x) for x in basis.row_status]
['At upper bound', 'At upper bound', 'At lower/fixed bound', 'At lower/fixed bound']
```

Perhaps @mdhaber or @mckib2 know more about why the choice was made to cast `primal_solution_status` (a `SolutionStatus` object) to a `HighsBasisStatus` object?

Note that, `SolutionStatus` objects are defined by:

```cpp
  py::enum_<SolutionStatus>(m, "SolutionStatus")
      .value("kSolutionStatusNone", SolutionStatus::kSolutionStatusNone)
      .value("kSolutionStatusInfeasible",
             SolutionStatus::kSolutionStatusInfeasible)
      .value("kSolutionStatusFeasible",
             SolutionStatus::kSolutionStatusFeasible);
```

While a `HighsBasisStatus` object contains:

```cpp
  py::enum_<HighsBasisStatus>(m, "HighsBasisStatus")
      .value("kLower", HighsBasisStatus::kLower)
      .value("kBasic", HighsBasisStatus::kBasic)
      .value("kUpper", HighsBasisStatus::kUpper)
      .value("kZero", HighsBasisStatus::kZero)
      .value("kNonbasic", HighsBasisStatus::kNonbasic);
```

mckib2 replied:
I agree, this looks like a bug as written.  There shouldn't be any reason to cast to a `HighsBasisStatus` type.

HaoZeke replied:
Awesome, thanks for verifying, resolving as fixed :)

HaoZeke commented:
This is ready pending the companion PR to SciPy `highs` (https://github.com/scipy/HiGHS/pull/63). Once that's in, the submodule and wrap file can point to it instead and this will be ready to go. (Getting it in upstream highs (https://github.com/ERGO-Code/HiGHS/pull/1460) would be good to have but isn't a blocker, and should happen in due time).

HaoZeke commented:
Once https://github.com/ERGO-Code/HiGHS/pull/1467 is in and https://github.com/scipy/HiGHS/pull/63 is updated, this will also close #15888.

HaoZeke commented:
@mtsokol is the numpy private namespace related error something being worked on elsewhere?

mtsokol commented:
> @mtsokol is the numpy private namespace related error something being worked on elsewhere?

Definitely caused by `numpy.core` rename. I will open a PR with a fix.

mtsokol commented:
@HaoZeke It looks that the error is coming from `array-api-compat` library vendored in SciPy. The fix for it was already merged four days ago here: https://github.com/data-apis/array-api-compat/pull/63

So I think `array-api-compat`s submodule needs to be checked-out to the latest commit.

HaoZeke commented:
> @HaoZeke It looks that the error is coming from `array-api-compat` library vendored in SciPy. The fix for it was already merged four days ago here: [data-apis/array-api-compat#63](https://github.com/data-apis/array-api-compat/pull/63).
> 
> So I think `array-api-compat`s submodule in SciPy just needs to be checked-out to the latest commit (by entering the submodule and checking-out to `main` branch pulling latest changes).

Awesome, that worked for the previous, but now there's one last [DeprecationWarning](https://github.com/scipy/scipy/actions/runs/6606360165/job/17942482362?pr=19255)..

```bash
DeprecationWarning: numpy.core.multiarray is deprecated and has been renamed to numpy._core.multiarray. The numpy._core namespace contains private NumPy internals and its use is discouraged, as NumPy internals can change without warning in any release. In practice, most real-world usage of numpy.core is to access functionality in the public NumPy API. If that is the case, use the public NumPy API. If not, you are using NumPy internals. If you would still like to access an internal attribute, use numpy._core.multiarray._ARRAY_API.
```

mtsokol commented:
Ok, this one will require a small PR to be fixed. 
In a nutshell - we already seen such errors for astropy as `_ARRAY_API` is imported from deprecated `numpy.core._multiarray_umath` from C level in objects built with NumPy 1.x. To solve it, deprecation warnings were removed for `_ARRAY_API` imports: https://github.com/numpy/numpy/pull/24943.

Here I see that the warning comes from the file `distance_pybind.cpp` in SciPy from pybind11 initialization `#include <pybind11/numpy.h>` (Here's a line in pybind header: https://github.com/pybind/pybind11/blob/3414c56b6c7c521d868c9a137ca2ace2e26b5b2e/include/pybind11/numpy.h#L283). 

I need to remove the same warning for `numpy.core.multiarray` - I will open it tomorrow!



https://github.com/numpy/numpy/pull/24943

mtsokol commented:
Here are two PRs to address warnings/errors present in dev CI stage:
numpy: https://github.com/numpy/numpy/pull/24985
scipy: https://github.com/scipy/scipy/pull/19426

HaoZeke commented:
@rgommers, with @mtsokol's timely PRs this should pass CI and I think it would be a good time to merge this in, there are still callbacks but they'd add significant complexity to the PR and should be reviewed separately. The [OOM bug](https://github.com/ERGO-Code/HiGHS/pull/1467) will require a fairly trivial update of the `.wrap` and submodule later.

h-vetinari commented:
I see a relevant failure in `Linux Meson tests / Prerelease deps, coverage, and 64-bit BLAS (3.9)`:
```
=========================== short test summary info ============================
FAILED scipy/optimize/tests/test_milp.py::test_milp_timeout_16545[options0-Time limit reached. (HiGHS Status 13:] - assert None is not None
= 1 failed, 57430 passed, 2497 skipped, 257 xfailed, 16 xpassed in 1535.82s (0:25:35) =
```

The refguide check gets killed - not 100% sure if this might be a flake, but other contemporary PRs pass that check.

h-vetinari commented on subprojects/highs.wrap:
```diff
@@ -0,0 +1,3 @@
+[wrap-git]
+url = https://github.com/scipy/highs.git
+revision = a0df06fb20f2c67c02cf84d8a3b72862c2ab2b27
```
Do we still need the wrap file, now that we have the full submodule? It shouldn't be, if I understand this [comment](https://github.com/scipy/scipy/pull/19255#issuecomment-1745669115) correctly.

HaoZeke replied:
We should provide both, especially in light of not providing the check in `meson.build` for this (it will use the `wrap` file automatically).

rgommers replied:
This works as intended it looks like; it basically checks out the git submodule automatically. The one minor issue is that it leaves behind an extra file:
```
subprojects/highs/.meson-subproject-wrap-hash.txt
```
If that's added to `.gitignore` then I think we're all good here.

rgommers replied:
That's not actually true it seems - when building an sdist after using the auto-checkout-via-wrapfile, I see:
```
WARNING: Submodule 'subprojects/highs' is not checked out and cannot be added to the dist
```

So you still need to run:
```
$ git submodule update --init
Submodule 'subprojects/highs' (https://github.com/scipy/highs) registered for path 'subprojects/highs'
```
That is then instant. But the annoying thing is that `git status` no longer tells you that you've missed a step. You've got the right checkout, but haven't "connected" it as a submodule. Nothing much goes wrong though as far as I can tell, it's just a little weird.

h-vetinari commented on .gitmodules:
```diff
@@ -9,14 +9,13 @@
 [submodule "scipy/_lib/array_api_compat"]
 	path = scipy/_lib/array_api_compat
 	url = https://github.com/data-apis/array-api-compat.git
+[submodule "subprojects/highs"]
```
Given that this is a break from our previous convention, could you please add a comment here to the rationale in the meson docs
```suggestion
# $REASON, see
# https://mesonbuild.com/Subprojects.html#why-must-all-subprojects-be-inside-a-single-directory
[submodule "subprojects/highs"]
```

h-vetinari commented on scipy/_lib/meson.build:
```diff
@@ -2,9 +2,6 @@ fs = import('fs')
 if not fs.exists('boost_math/README.md')
   error('Missing the `boost` submodule! Run `git submodule update --init` to fix this.')
 endif
-if not fs.exists('highs/README.md')
-  error('Missing the `highs` submodule! Run `git submodule update --init` to fix this.')
-endif
```
Can you replicate that check in `scipy/optimize/_highs/meson.build` (or where appropriate)?

HaoZeke replied:
Not required since `meson` will "do the right thing" with the `wrap` file if the folder is missing.

h-vetinari commented on scipy/optimize/_linprog_highs.py:
```diff
@@ -338,24 +426,23 @@ def _linprog_highs(lp, solver, time_limit=None, presolve=True,
     options = {
-        'presolve': presolve,
-        'sense': HIGHS_OBJECTIVE_SENSE_MINIMIZE,
-        'solver': solver,
```
I'd prefer to keep the `black`-style formatting changes out of this PR, as it makes it much harder to follow what changed in a meaningful way, and what just got some formatting changes.

HaoZeke replied:
Taking another look at the `.pyx` file there's really almost nothing which gets brought forward, so I'm not sure what the goal would be, the file is both much smaller and logically restructured as well... Is there a particular part of the `pyx` file which seems to be relevant? This would basically require reworking the PR from scratch for what would seem to be very little gain, a full rewrite to `highspy` is IMO the right time to switch formatting as well...

HaoZeke replied:
> This would basically require reworking the PR from scratch...

By which I mean restoring the `pyx` file, renaming it, and then deleting / rewriting in a series of new commits.

HaoZeke replied:
I tried this actually, and it will still show up as deleted in the final (consolidated) `git` diff, not moved (even though the earlier commits now [show a rename](https://github.com/scipy/scipy/pull/19255/commits/b19e74d5a723d07c114309d949f8137f1fcfee1b)).

h-vetinari reviewed (commented):
Drive-by review

h-vetinari commented on scipy/optimize/_highs/cython/src/_highs_wrapper.pyx:
```diff
@@ -1,736 +0,0 @@
-        double[::1] ub,
-        np.uint8_t[::1] integrality,
-        dict options):
-    '''Solve linear programs using HiGHS [1]_.
-
-    Assume problems of the form:
-
-        MIN c.T @ x
-        s.t. lhs <= A @ x <= rhs
-             lb <= x <= ub
-
-    Parameters
-    ----------
```
Please replicate this docstring in the new highs-wrapper (adapted if/where necessary)

mckib2 replied:
I suspect parts of this docstring were out of date to start with as it was written at the start but forgotten about during several updates (as it's not actually rendered anywhere).  If there are interesting bits for the wrapper specifically (such as undocumented HiGHS developer options), those would be the things to pull forward

HaoZeke replied:
Given that it is not rendered / verified, would it still make sense to try to port it? @h-vetinari

h-vetinari replied:
> Given that it is not rendered / verified, would it still make sense to try to port it? @h-vetinari

If it's not used and the actual docstring comes from somewhere else, then I agree that it's best to delete the redundant/outdated one.

mckib2 commented:
> I see a relevant failure in `Linux Meson tests / Prerelease deps, coverage, and 64-bit BLAS (3.9)`:
> 
> ```
> =========================== short test summary info ============================
> FAILED scipy/optimize/tests/test_milp.py::test_milp_timeout_16545[options0-Time limit reached. (HiGHS Status 13:] - assert None is not None
> = 1 failed, 57430 passed, 2497 skipped, 257 xfailed, 16 xpassed in 1535.82s (0:25:35) =
> ```
> 
> The refguide check gets killed - not 100% sure if this might be a flake, but other contemporary PRs pass that check.

I believe this is a known flaky test (e.g., gh-18339).  Sometimes it finds a solution and quits before exhaustion, and sometimes it doesn't.  The "right thing" to do here IMO is to disable the test until MIP callbacks are implemented, remove the timeout, and check that the solution is preserved after hitting the first callback announcing a solution has been found.

If you want to keep the test, messing with the timeout may be necessary as upstream fixes/updates will inevitably change the window of opportunity for seeing a solution and quitting before done.

h-vetinari commented:
OK, thanks for trying to see if git manages to detect the file rename. This was mainly me asking because merging 45 commits with a non-negligible amount of churn would normally be squash-merged.

That said, if you're up to doing an interactive rebase that condenses the history to 2-5 clean commits with self-contained increments (e.g. one could be centered around the file rename), that would actually be the best solution IMO.

lucascolley commented:
> Test failure doesn't seem to be related..

yep, seeing that on other PRs

h-vetinari commented:
Thanks for the clean-up here @HaoZeke!

I'll admit that I find this hard to review (there's still a fair amount of back and forth in the commits, there are unrelated formatting changes mixed in, and few indications what got "just" translated rather than written from scratch, etc.), but given that this seems to pass the existing tests, that's a very good sign already (assuming we have some halfway decent test coverage 😅).

HaoZeke commented:
@h-vetinari I could perhaps do another "from scratch" clean-room implementation in a few weeks, which would fix (at-least) the formatting, but perhaps it might be more expedient to let it out into the wild as is? I'd be happy to make other changes to get this up to speed.

rgommers commented:
> but perhaps it might be more expedient to let it out into the wild as is

+1 for this. I haven't forgotten about this PR, it was just a busy period + holiday season. Planning to get this reviewed this week.

rgommers commented on scipy/optimize/_linprog_highs.py:
```diff
@@ -13,108 +13,197 @@
-        return choices[default_str]
-
-
-def _linprog_highs(lp, solver, time_limit=None, presolve=True,
```
A large majority of lines in the diff from here on down in this file are style chances (single to double quotes, etc.). It's now hard to disentangle that, but for future reference: please avoid mixing in style changes, especially very minor ones that don't really improve readability - it makes reviewing harder.

HaoZeke replied:
Yup, sorry about that, I reversed the formatting changes so it should be clearer now.

rgommers reviewed (commented):
This is very close. A few comments inline - all minor.

The silencing of unused variable warnings didn't work, because they're coming from the static library built within `subprojects/highs`:
```
[6/11] Compiling C++ object subprojects/highs/src/libhighs.a.p/lp_data_HighsSolve.cpp.o
../subprojects/highs/src/lp_data/HighsSolve.cpp:48:10: warning: unused variable 'imprecise_solution' [-Wunused-variable]
    bool imprecise_solution;
         ^
../subprojects/highs/src/l
```

Looks like those should be fixed upstream, and perhaps silenced in our forked repo in the meantime.

There are also a couple of other warnings. In the build:
```
ld: warning: ignoring duplicate libraries: '-lc++'
```
and in the configure stage:
```
highs| subprojects/highs/meson.build:16: WARNING: add_languages is missing native:, assuming languages are wanted for both host and build.
```
I'm seeing a number of other issues in `highs/meson.build` (mainly with compiler-specific flags that aren't check) that are likely to be a problem, so let me open a PR for those upstream.

I'll push a rebase to fix the conflict and get a fresh CI run.

rgommers commented:
The linter is unhappy - the `No explicit `stacklevel` keyword argument found` at least look like they need to be fixed up.

HaoZeke commented:
> > but perhaps it might be more expedient to let it out into the wild as is
> 
> +1 for this. I haven't forgotten about this PR, it was just a busy period + holiday season. Planning to get this reviewed this week.

Thanks for the comprehensive review and comments @rgommers, I'll address these both upstream and here this weekend.

HaoZeke commented:
@rgommers I've addressed the issues raised on a PR to HiGHS here:
https://github.com/ERGO-Code/HiGHS/pull/1602

Beyond that I've made a few cosmetic changes (rolling back `black`) and I think I've responded to all the comments (below):

- [X] `minimum Meson version is set to 1.2.0`
	- Set to match SciPy's `1.1.0`
- [x] `optional dependency on `zlib`; looks like that is for file reading code only and we should disable it unconditionally when calling `subproject('highs')``
- [X] `usage of `git` for a version string in `highspy` (looks robust to failure, but still)`
	- Not used if called as subproject
- [X] `a `run_command('python3', '-c', ...` that looks less robust`
	- Not used if called as a subproject
- [ ] `logic for `threads` and `libatomic` dependencies that now needs to be kept in sync in two places (our own versions, and the copied ones in highs)`
    - [ ] `has`threads_dep = dependency('threads', required: true` while we have `required: false`. I'm not sure if it can fail on niche platforms (mingw is problematic IIRC).`
- I'm not sure what can be done for this one, `threads` are a dependency of `highs`, and the `libatomic` fallback is required, not just for SciPy, so I've left it in
- [X] `An unnecessary dependency on `libm` (our top-level `meson.build` explains why it's only needed for C)`
	- Only used if the `C` interface is built
- [X] `an explicit `declare_dependency(link_args: '-lstdc++')` that doesn't look right`
	- Removed

- [X] The linter is unhappy - the `No explicit stacklevel keyword argument found` at least look like they need to be fixed up.
	- The `linter` errors  seems spurious, all the `warn` calls have `stacklevel` set explicitly, and the line flagged isn't that long
	- Can't reproduce locally with `ruff lint scipy/optimize | grep highs`

HaoZeke commented:
[Failing tests](https://github.com/scipy/scipy/actions/runs/7769105154/job/21187811506?pr=19255) seem unrelated:
```
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_method[uint8] - OverflowError: Python integer 504 out of bounds for uint8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_method[int8] - OverflowError: Python integer 184 out of bounds for int8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_rank3_valid[uint8] - OverflowError: Python integer 504 out of bounds for uint8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_rank3_valid[int8] - OverflowError: Python integer 184 out of bounds for int8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_rank3_same[uint8] - OverflowError: Python integer 504 out of bounds for uint8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_rank3_same[int8] - OverflowError: Python integer 184 out of bounds for int8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_rank3_all[uint8] - OverflowError: Python integer 504 out of bounds for uint8
FAILED scipy/signal/tests/test_signaltools.py::TestCorrelateReal::test_rank3_all[int8] - OverflowError: Python integer 184 out of bounds for int8
= 8 failed, 45923 passed, 2378 skipped, 147 xfailed, 16 xpassed in 342.04s (0:05:42) =
Error: Process completed with exit code 1.
```

lucascolley commented on tools/lint.toml:
```diff
@@ -5,16 +5,16 @@ line-length = 88
 # Enable Pyflakes `E` and `F` and PyUpgrade `UP` codes by default.
 # Also, `PGH004` which checks for blanket (non-specific) `noqa`s
 # and `B028` which checks that warnings include the `stacklevel` keyword.
 # `B028` added in gh-19623.
-select = ["E", "F", "PGH004", "UP", "B028"]
-ignore = ["E741"]
+lint.select = ["E", "F", "PGH004", "UP", "B028"]
+lint.ignore = ["E741"]
+lint.exclude = ["scipy/datasets/_registry.py"]
 
 # Allow unused variables when underscore-prefixed.
-dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
+lint.dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
```
is this necessary?

HaoZeke replied:
Nope, good point, removed.

tylerjereddy commented:
@HaoZeke I'd like to branch for SciPy `1.13.0` around Sunday (17th) -- will/should this be ready by then or should we bump the milestone to `1.14.0`?

rgommers commented:
I'd much prefer to merge this right after branching `1.13.x`. This doesn't seem critical for NumPy 2.0 support, and it's too large a change to merge at the last minute.

jajhall commented:
Part of the problem exposed in

https://github.com/ERGO-Code/HiGHS/issues/1693

is that Google is not using the latest version of SciPy, so can't update HiGHS beyond v1.4.1

@galabovaa has `highspy` working in

https://pypi.org/project/highspy/1.7.1.dev1/

and I have a few follow-on issues from releasing v1.7.0 to deal with before we release HiGHS/highspy 1.7.1

rgommers commented:
Thanks for sharing @jajhall. Let's make sure we get this merged soon then - it will make it much easier to upgrade and unvendor HiGHS.

h-vetinari commented:
Any update from the highs side? 1.14 is going to come around pretty soon, it would be good not to let this slip that release.

HaoZeke commented:
AFAIK this isn't waiting on HiGHs, and should be ready for review.

On 4/16/24 11:48 PM, h-vetinari ***@***.***> wrote:
> Any update from the highs side? 1.14 is going to come around pretty 
> soon, it would be good not to let this slip that release.
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/scipy/scipy/pull/19255#issuecomment-2060082136>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABBCUTYASTF6SQ6JSVYPGI3Y5W2FXAVCNFSM6AAAAAA437DDQOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANRQGA4DEMJTGY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
> 
>

h-vetinari commented:
Really? I interpreted

> I have a few follow-on issues from releasing v1.7.0 to deal with before we release HiGHS/highspy 1.7.1

as requiring us to wait on the 1.7.1 release, which hasn't happened yet AFAICT.

jajhall commented:
I don't know what methods in 'highspy' are used by Scipy, but it is unlikely to be affected by the issues I mentioned

rgommers commented:
Doing final due diligence on this. It looks quite good. Some new build warnings, but way more are removed so this is a net win. The HiGHS build is very heavy, and the changes somehow moved the build parallelism in the wrong direction:

<img width="1047" alt="image" src="https://github.com/scipy/scipy/assets/98330/d5dd9da2-bef8-4d7e-b2d0-a2da34d37e0b">

On `main` it's usually `sparsetools` that finishes last, because that's the heaviest component by some distance.

I verified that the total contribution to build time from HiGHS didn't change much - it's about 25 seconds on a 12-core CPU on both `main` and this branch:

<img width="1047" alt="image" src="https://github.com/scipy/scipy/assets/98330/a782ed5a-bb62-4a6c-932c-91018648c5db">

I'll try reshuffling this a bit.

rgommers commented:
The last commit should fix it. The `_highs.so` target still finishes last, but not much later than other targets anymore:

<img width="872" alt="image" src="https://github.com/scipy/scipy/assets/98330/72b244d4-fce4-45cd-9377-feb8e525176c">

eli-schwartz commented:
Why does this matter? Are you just causing the generated targets to textually move around inside build.ninja to tickle the job scheduler into running some targets before others when you know they take a long time?

Seems unreliable :p note that ninja schedules jobs based on the pointer value of its internal data structure, and the first job to run depends on the specifics of how `malloc()` returns values.

If it works, it... works? I wouldn't pin my hopes on it though.

rgommers commented:
> Why does this matter? Are you just causing the generated targets to textually move around inside build.ninja to tickle the job scheduler into running some targets before others when you know they take a long time?

Exactly, yes. I don't think it's unreliable? If you have a bunch of `subdir()` statements, the order matters. If you put the subdir containing the target that takes the longest to build at the end, the parallelism ends up being very unbalanced. AFAIK this comment is correct (at least in practice, not sure if `ninja` could in theory still reorder things):

https://github.com/scipy/scipy/blob/bd60b4ef9d886a9171345fc064c80aad6d171e73/scipy/meson.build#L527-L559

eli-schwartz commented:
> AFAIK this comment is correct (at least in practice, not sure if `ninja` could in theory still reorder things):

I think the answer is "the comment is correct, that is indeed what ninja does, but you're depending on undocumented internals".

Samurai, the c99 reimplementation of ninja used by e.g. Alpine Linux, uses a stack which means "last scheduled first executed", breaking your assumption. Admittedly, samurai also doesn't yet implement Fortran dyndeps, so you'd be well within your rights to not care in the slightest.

rgommers commented:
Thanks for the context @eli-schwartz, that is good to know. I'll see about checking how things look with Samurai too, to ensure it's not completely terrible. I think all our CI and >95% of users use `ninja`, so we should optimize for that mostly. But I'd like Samurai to not be a lot worse at least.

rgommers commented:
I checked the benchmark locally, it didn't change. It's just very slow and close to the timeout limit it looks like. On this branch:
```
[25.00%] ··· optimize_milp.MilpMagicSquare.time_magic_square                                    ok
[25.00%] ··· ====== ==========
              size            
             ------ ----------
               3     25.3±0ms 
               4     1.13±0s  
               5     8.50±0s  
             ====== ==========
```

On `main`:
```
[25.00%] ··· optimize_milp.MilpMagicSquare.time_magic_square                                    ok
[25.00%] ··· ====== ==========
              size            
             ------ ----------
               3     26.2±0ms 
               4     1.11±0s  
               5     8.35±0s  
             ====== ==========
```

rgommers commented:
Okay this benchmark is just way too slow and there is already an issue for it: gh-19389. Disabling it now, because it takes over a minute to run for that single benchmark locally as well.

mdhaber commented:
Thanks @HaoZeke @rgommers @h-vetinari! Is scipy/HiGHS at the commit we want it to be at? There were some bug reports that were fixed upstream; it would be good to make sure they're resolved here.

rgommers commented:
> Is scipy/HiGHS at the commit we want it to be at (and does SciPy point to the right commit)?

`subprojects/highs/Version.txt` says that we're now at `1.6.0`. `scipy/HiGHS` cannot be at the wrong commit really, it's easiest to just look what you have on disk locally after checking out the submodule.

We should bump it further now that the hard work is done. I was planning to submit a follow-up PR to HiGHS first to fix some warnings and loose ends.

> `python3` is not a command in my conda environment. I can create an alias, I'm sure, but I'm not sure if this should be necessary?

That's a bug indeed, that code should not be trying to invoke `python3`. I'll fix that up right away since that broke the build for you (and I'm sure for a few other people).

rgommers commented:
> and does SciPy point to the right commit

Ah, what did happen is that @HaoZeke made a few fixes upstream to address my review comments and those are not yet pulled into `scipy/highs`. Doesn't include the `python3` thing though. A CI job for conda on Windows would have caught that - hopefully we'll get that at some point, we need it for checking Flang as well.

rgommers commented:
Okay, I'm going to have to revert this. All my critical build comments in https://github.com/scipy/scipy/pull/19255#issuecomment-1925424141 were addressed upstream in https://github.com/ERGO-Code/HiGHS/pull/1603, however that PR was not included in `scipy/highs`.

It turns out that it's not easy to update the submodule from 1.6.0 to 1.7.0, because the Meson build got broken and CI disabled in the process of the 1.7.0 work in Feb/Mar. It's a lot harder to now retroactively figure out what went wrong, compared to leaving CI in place and fixing things as they come up. 

There are also a lot of other little issues. I opened https://github.com/ERGO-Code/HiGHS/pull/1737 to address them, but there's a bit more.

One builtin test (`test_lpsolvers`) is failing with a UBSAN error. And, most importantly, `highspy` builds but does not import.

Worst case we can apply all the needed patches to 1.6.0 in `scipy/highs` and then re-land this PR. However, it'd be much better to fix up the `latest` branch upstream and update to at least 1.7.0 in one go. Otherwise the longer we wait, the more it goes out of date again.

HaoZeke commented:
Yikes, sorry I didn't notice the activity on this PR, and that I didn't update the scipy fork of highs before this went it. Will have a an updated PR ready soon.

On 4/26/24 9:01 PM, Ralf Gommers ***@***.***> wrote:
> Okay, I'm going to have to revert this. All my critical build comments 
> in #19255 (comment) 
> <https://github.com/scipy/scipy/pull/19255#issuecomment-1925424141> were 
> addressed upstream in ERGO-Code/HiGHS#1603 
> <https://github.com/ERGO-Code/HiGHS/pull/1603>, however that PR was not 
> included in |scipy/highs|.
> 
> It turns out that it's not easy to update the submodule from 1.6.0 to 
> 1.7.0, because the Meson build got broken and CI disabled in the process 
> of the 1.7.0 work in Feb/Mar. It's a lot harder to now retroactively 
> figure out what went wrong, compared to leaving CI in place and fixing 
> things as they come up.
> 
> There are also a lot of other little issues. I opened 
> ERGO-Code/HiGHS#1737 <https://github.com/ERGO-Code/HiGHS/pull/1737> to 
> address them, but there's a bit more.
> 
> One builtin test (|test_lpsolvers|) is failing with a UBSAN error. And, 
> most importantly, |highspy| builds but does not import.
> 
> Worst case we can apply all the needed patches to 1.6.0 in |scipy/highs| 
> and then re-land this PR. However, it'd be much better to fix up the 
> |latest| branch upstream and update to at least 1.7.0 in one go. 
> Otherwise the longer we wait, the more it goes out of date again.
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/scipy/scipy/pull/19255#issuecomment-2080113574>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABBCUTYVK2OYULI5QWHQM3DY7K6DPAVCNFSM6AAAAAA437DDQOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAOBQGEYTGNJXGQ>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
> 
>

h-vetinari commented:
> Will have a an updated PR ready soon.

@HaoZeke, any updates on this? 🙃
